earth2grid/healpix.py

```diff
@@ -602,3 +599,154 @@ def conv2d(input, weight, bias=None, stride=1, padding=0, dilation=1, groups=1):
     weight = weight.unsqueeze(-3)
     out = torch.nn.functional.conv3d(input, weight, bias, stride, padding, dilation, groups)
     return einops.rearrange(out, "n c f x y -> n c () (f x y)")
+
+
+def _pixels_to_rings(nside: int, p: ArrayT) -> ArrayT:
+    """Get the ring number of a pixel ``i`` in RING order"""
+    # See eq (2-5) of Gorski
+    xp = _get_array_library(p)
+    npix = 12 * nside * nside
+    ncap = 2 * nside * (nside - 1)
+
+    i_north = xp.floor(0.5 * (1 + np.sqrt(1 + 2 * p)))
+    j_north = p - 2 * (i_north - 1) * i_north
+
+    p_eq = p - ncap
+    i_eq = xp.floor(p_eq / (4 * nside)) + nside - 1
+    j_eq = p_eq % (4 * nside)
+
+    p_south = npix - p - 1
+    i_south = xp.floor(0.5 * (1 + np.sqrt(1 + 2 * p_south)))
+    j_south = p_south - 2 * (i_south - 1) * i_south
+    length_south = i_south * 4
+
+    i = i_north - 1
+    i = xp.where(p >= ncap, i_eq, i)
+    i = xp.where(p >= (npix - ncap), 4 * nside - i_south - 1, i)
+
+    j = j_north
+    j = xp.where(p >= ncap, j_eq, j)
+    j = xp.where(p >= (npix - ncap), length_south - 1 - j_south, j)
+
+    return i.astype(int), j.astype(int)
+
+
+def ring_length(nside: int, i: ArrayT) -> ArrayT:
+    """The number of pixels in ring 0 <= i < 4 * nside - 2"""
+    xp = _get_array_library(i)
+
+    length_north = 4 * (i + 1)
+    length_eq = 4 * nside
+    length_south = (4 * nside - i - 1) * 4
```

Comment:
This seems like a reimplementation of the [healpix_bare utilities](https://github.com/NVlabs/earth2grid/blob/main/earth2grid/third_party/healpix_bare/healpix_bare.c#L179), is this mainly to add GPU support?

Reply:
oh. i didn't realize they had these things easily available. I recoded this by hand from reading the gorski paper.